tket/src/Circuit/Circuit.cpp

```diff
@@ -320,4 +328,104 @@ bool Circuit::in_causal_order(
+                            [](const ResourceData& a, const ResourceData& b) {
+                              return a.GateDepth.min < b.GateDepth.min;
+                            })
+                            ->GateDepth.min;
+  data.GateDepth.max += std::max_element(
+                            pre_data.begin(), pre_data.end(),
+                            [](const ResourceData& a, const ResourceData& b) {
+                              return a.GateDepth.max < b.GateDepth.max;
+                            })
+                            ->GateDepth.max;
+  // 2. OpTypeDepth
+  std::map<OpType, unsigned> min_depths;
+  std::map<OpType, unsigned> max_depths;
+  for (const ResourceData& pre_v_data : pre_data) {
+    for (const auto& pair : pre_v_data.OpTypeDepth) {
+      if (pair.second.min > min_depths[pair.first]) {
+        min_depths[pair.first] = pair.second.min;
+      }
+      if (pair.second.max > max_depths[pair.first]) {
+        max_depths[pair.first] = pair.second.max;
+      }
+    }
+  }
+  for (const auto& pair : min_depths) {
+    data.OpTypeDepth[pair.first].min += min_depths[pair.first];
+  }
+  for (const auto& pair : max_depths) {
+    data.OpTypeDepth[pair.first].max += max_depths[pair.first];
+  }
+  // 3. TwoQubitGateDepth
+  data.TwoQubitGateDepth.min +=
+      std::max_element(
+          pre_data.begin(), pre_data.end(),
+          [](const ResourceData& a, const ResourceData& b) {
+            return a.TwoQubitGateDepth.min < b.TwoQubitGateDepth.min;
+          })
+          ->TwoQubitGateDepth.min;
+  data.TwoQubitGateDepth.max +=
+      std::max_element(
+          pre_data.begin(), pre_data.end(),
+          [](const ResourceData& a, const ResourceData& b) {
+            return a.TwoQubitGateDepth.max < b.TwoQubitGateDepth.max;
+          })
+          ->TwoQubitGateDepth.max;
+}
+
+ResourceData Circuit::get_resources() /*const*/ {
+  // Traverse the DAG in topological order. At each vertex compute a new
+  // ResourceData based on the ResourceData already computed for its immediate
+  // predecessors. Compute final ResourceData based on terminal nodes.
+  const VertexVec vertices = vertices_in_order();
+  std::map<Vertex, ResourceData> datamap;
+  std::map<OpType, ResourceBounds<unsigned>> op_type_count;
+  for (const Vertex& v : vertices) {
+    ResourceData data;
+    OpType optype = get_OpType_from_Vertex(v);
+    if (!is_initial_type(optype)) {
+      if (!is_final_type(optype)) {
+        if (optype == OpType::DummyBox) {
+          const DummyBox& dbox =
+              static_cast<const DummyBox&>(*get_Op_ptr_from_Vertex(v));
+          data = dbox.get_resource_data();
+          for (const auto& pair : data.OpTypeCount) {
+            op_type_count[pair.first].min += pair.second.min;
+            op_type_count[pair.first].max += pair.second.max;
+          }
+        } else {
+          data.GateDepth = {1};
+          data.OpTypeDepth[optype] = {1};
+          if (OpDesc(optype).n_qubits() == 2) {
```

Comment:
How should we treat types with dynamic n_qubits? e.g. CnRy, boxes.

Reply:
Changed to check for actual number of qubits.